extensions/applicationinsights-debugplugin-js/src/components/helpers.ts

```diff
@@ -0,0 +1,179 @@
+import { CoreUtils } from '@microsoft/applicationinsights-core-js';
+
+const MAX_DEPTH = 8;
+
+export const isArray = (obj: any): obj is any[] => {
+  return Object.prototype.toString.call(obj) === '[object Array]';
+}
+
+const traverseAndReplace = (target: Object, maxDepth: number, currentDepth: number, thingsReferenced?: string[]): Object => {
+  const out = {};
+
+  if (!thingsReferenced) { thingsReferenced = []; }
+  if (CoreUtils.isTypeof(target, "object")) {
+    for (const key of CoreUtils.objKeys(target)) {
+      if (CoreUtils.arrIndexOf(thingsReferenced, target[key]) !== -1) {
+        out[key] = `<circular (${key})>`;
+      }
+      else if (target[key] !== null && CoreUtils.isTypeof(target[key], 'object')) {
+        if (currentDepth >= maxDepth) {
+          out[key] = '<max allowed depth reached>';
+        } else {
+          thingsReferenced.push(target[key]);
+          out[key] = traverseAndReplace(target[key], maxDepth, currentDepth + 1, thingsReferenced);
+          thingsReferenced.pop();
+        }
+      }
+      else {
+        out[key] = target[key];
+      }
+    }
+  }
+
+  return out;
+}
+
+export class LoggerElement {
+  public lastSelectedElement: HTMLElement;
+  public lastSelectedObject: Object;
+
+  public el: HTMLElement;
+
+  constructor(parent: HTMLElement) {
+    const s = this;
+    s.el = document.createElement("div");
+    s.el.className = 'my-logger';
+
+    const copyButton = document.createElement("button");
+    copyButton.innerText = "copy current node";
```

Comment:
Set innerText to constant string?

Reply:
yep! That's just how buttons work in HTML. `<button>text goes here</button>`